### PR TITLE
Implement/consensus/rpc #16

### PR DESCRIPTION
**Pull request by x-senpai-x**
Implemented consensus_rpc.go, nimbus_rpc.go,mock_rpc.go .In mock_rpc.go used the testdata of helios  and added it here in testdata folder , made a folder consensus/consensus_core to define the structs

### CODE REVIEW

**Comment by x-senpai-x** [comment]
@star-gazer111  please review this

File: consensus/consensus_core/consensus_core.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+type SignedBlsToExecutionChange struct {
+	message   BlsToExecutionChange
+	signature SignatureBytes
+}
+type BlsToExecutionChange struct {
+	validator_index uint64
+	from_bls_pubkey [48]byte
+}
+type BeaconBlockBody struct { //not implemented
+	randao_reveal            SignatureBytes
```
Comment (gerceboss) [inline consensus/consensus_core/consensus_core.go:120]: correct the spelling mistakes please, check at other places as well

Comment (gerceboss) [reply consensus/consensus_core/consensus_core.go:120]: Add unit tests as well for some functions that use concurrency

**Comment by gerceboss** [comment]
I think we need to split this PR in two parts, one which adds data(json) and other adding the code logic. What's your take on this ? @star-gazer111

File: consensus/consensus_core/consensus_core.go (hunk 1 above)
Comment (x-senpai-x) [reply consensus/consensus_core/consensus_core.go:120]: > correct the spelling mistakes please, check at other places as well

@gerceboss Sir which spelling mistakes ? ig the only spelling mistakes I have made are in comments that I have added other than that for naming conventions at most places I have copied the names from helios

**Comment by star-gazer111** [comment]
> I think we need to split this PR in two parts, one which adds data(json) and other adding the code logic. What's your take on this ? @star-gazer111

I think we can go ahead with this only for now.

File: common/types.go [hunk 2]
```diff
@@ -2,107 +2,109 @@ package common
...
-   if len(t.Hashes)>0{
-      return json.Marshal(t.Hashes)
-   }
-   return json.Marshal(t.Full)
+func (t Transactions) MarshalJSON() ([]byte, error) {
+	if len(t.Hashes) > 0 {
+		return json.Marshal(t.Hashes)
+	}
+	return json.Marshal(t.Full)
 }
```
Comment (star-gazer111) [inline common/types.go:71]: What did you change here?

File: common/types.go [hunk 3]
```diff
@@ -2,107 +2,109 @@ package common
...
 
 // can be an enum
 type BlockTag struct {
-   Latest bool
-   Finalized bool
-   Number uint64
+	Latest    bool
+	Finalized bool
+	Number    uint64
 }
```
Comment (star-gazer111) [inline common/types.go:79]: and here?

File: common/types.go [hunk 4]
```diff
@@ -2,107 +2,109 @@ package common
...
-      return "finalized"
-   }
-   return fmt.Sprintf("%d", b.Number)
+	if b.Latest {
+		return "latest"
+	}
+	if b.Finalized {
+		return "finalized"
+	}
+	return fmt.Sprintf("%d", b.Number)
```
Comment (star-gazer111) [inline common/types.go:87]: here also?

File: common/types.go [hunk 5]
```diff
@@ -2,107 +2,109 @@ package common
...
+	case "latest":
+		b.Latest = true
+	case "finalized":
+		b.Finalized = true
+	default:
+		var err error
+		b.Number, err = parseBlockNumber(block)
+		if err != nil {
+			return err
+		}
```
Comment (star-gazer111) [inline common/types.go:105]: these extra whitespaces are for?

File: consensus/rpc/nimbus_rpc.go [hunk 6]
```diff
@@ -1,27 +1,141 @@
...
+	"strconv"
+	"time"
+)
+
 // uses types package
+const (
+	MAX_REQUEST_LIGHT_CLIENT_UPDATES uint8 = 128
+	initialBackoff                         = 50 * time.Millisecond // Initial backoff of 50ms
+	maxRetries                             = 5                     // Number of retry attempts
+)
```
Comment (star-gazer111) [inline consensus/rpc/nimbus_rpc.go:18]: are these values taken from helios ? @x-senpai-x

**Review by star-gazer111 — CHANGES_REQUESTED** [review]

File: common/types.go (hunk 2 above)
Comment (x-senpai-x) [reply common/types.go:71]: > What did you change here?
Before issuing the first pr i didn't use gofmt 
So I ran it before issuing the second pr 
But There have been no changes in this file

File: common/types.go (hunk 5 above)
Comment (x-senpai-x) [reply common/types.go:105]: In common/types.go whitespaces were introduced after running gofmt

File: consensus/rpc/nimbus_rpc.go (hunk 6 above)
Comment (x-senpai-x) [reply consensus/rpc/nimbus_rpc.go:18]: in helios they used default settings of backoffsettings package which had  the same values

**Review by star-gazer111 — APPROVED** [review]
LGTM.

File: consensus/consensus_core/consensus_core.go (hunk 1 above)
Comment (star-gazer111) [reply consensus/consensus_core/consensus_core.go:120]: We can mereg this once @gerceboss approves at his end as well.

**Comment by gerceboss** [comment]
LGTM as well, merging the  @x-senpai-x , just mention while making the PR if you are adding any constant values referencing to helios

**Pull request merged**